File: internal/app/compact_command.go
package app

import (
	"fmt"
	"strconv"
	"strings"
)

const compactThresholdMetadataKey = "context_compaction_threshold_percent"

type compactCommandOptions struct {
	Note             string
	ThresholdPercent float64
	HasThreshold     bool
}

func parseCompactCommandArgs(args []string) compactCommandOptions {
	tokens := make([]string, 0, len(args))
	for _, arg := range args {
		trimmed := strings.TrimSpace(arg)
		if trimmed != "" {
			tokens = append(tokens, trimmed)
		}
	}
	if len(tokens) == 0 {
		return compactCommandOptions{}
	}

	parsed := compactCommandOptions{}
	switch {
	case len(tokens) >= 1:
		if threshold, ok := parseCompactThresholdToken(tokens[0]); ok {
			parsed.ThresholdPercent = threshold
			parsed.HasThreshold = true
			tokens = tokens[1:]
			break
		}
		fallthrough
	case len(tokens) >= 2:
		if strings.EqualFold(tokens[0], "threshold") {
			if threshold, ok := parseCompactThresholdToken(tokens[1]); ok {
				parsed.ThresholdPercent = threshold
				parsed.HasThreshold = true
				tokens = tokens[2:]
			}
		}
	}
	parsed.Note = strings.TrimSpace(strings.Join(tokens, " "))
	return parsed
}

func parseCompactThresholdToken(token string) (float64, bool) {
	token = strings.TrimSpace(token)
	if !strings.HasSuffix(token, "%") {
		return 0, false
	}
	raw := strings.TrimSpace(strings.TrimSuffix(token, "%"))
	if raw == "" {
		return 0, false
	}
	value, err := strconv.ParseFloat(raw, 64)
	if err != nil {
		return 0, false
	}
	if value < 0 {
		value = 0
	}
	if value > 100 {
		value = 100
	}
	return value, true
}

func formatCompactThresholdPercent(value float64) string {
	value = normalizeCompactThresholdPercent(value)
	if value == float64(int64(value)) {
		return fmt.Sprintf("%.0f%%", value)
	}
	return fmt.Sprintf("%.1f%%", value)
}

func normalizeCompactThresholdPercent(value float64) float64 {
	if value < 0 {
		return 0
	}
	if value > 100 {
		return 100
	}
	return value
}

File: web/src/features/desktop/chat/services/compact-command.ts
export const COMPACT_THRESHOLD_METADATA_KEY = 'context_compaction_threshold_percent'

export interface ParsedCompactCommand {
  note: string
  hasThreshold: boolean
  thresholdPercent: number
}

function parseCompactThresholdToken(token: string): number | null {
  const trimmed = token.trim()
  if (!trimmed.endsWith('%')) {
    return null
  }
  const raw = trimmed.slice(0, -1).trim()
  if (!raw) {
    return null
  }
  const parsed = Number.parseFloat(raw)
  if (!Number.isFinite(parsed)) {
    return null
  }
  if (parsed <= 0) {
    return 0
  }
  if (parsed >= 100) {
    return 100
  }
  return parsed
}

export function parseCompactCommandInput(input: string): ParsedCompactCommand {
  const trimmed = input.trim()
  const compactMatch = trimmed.match(/^\/compact(?:\s+(.*))?$/i)
  const tail = compactMatch?.[1]?.trim() ?? ''
  if (!tail) {
    return {
      note: '',
      hasThreshold: false,
      thresholdPercent: 0,
    }
  }

  const tokens = tail.split(/\s+/).filter(Boolean)
  let thresholdPercent = 0
  let hasThreshold = false

  if (tokens.length >= 1) {
    const parsed = parseCompactThresholdToken(tokens[0])
    if (parsed !== null) {
      thresholdPercent = parsed
      hasThreshold = true
      tokens.splice(0, 1)
    }
  }
  if (!hasThreshold && tokens.length >= 2 && tokens[0].toLowerCase() === 'threshold') {
    const parsed = parseCompactThresholdToken(tokens[1])
    if (parsed !== null) {
      thresholdPercent = parsed
      hasThreshold = true
      tokens.splice(0, 2)
    }
  }

  return {
    note: tokens.join(' ').trim(),
    hasThreshold,
    thresholdPercent,
  }
}
